LOAD
<cghelptext>
<div style="font-size=25px; text-align:justify">
Select &quot;<b>Own data&quot;</b> to load your data in <b>SPSS</b>, <b>Excel</b>, <b>R</b> or <b>plain text</b> format from a file. <br>
Once the data is loaded satisfactory, click on &quot;OK&quot; button and the second panel (&quot;Step 2. Select variables&quot;) will open automatically.<br><br>
Also you can select example data already present in the <a href=\"http://cran.r-project.org/web/packages/compareGroups/compareGroups.pdf\"><span style=\"font-family:courier new,courier,monospace;\"><strong>compareGroups</strong></span></a> package (<b>REGICOR</b> and <b>PREDIMED</b>) and <b>SNPs</b> from <a href=\"http://cran.r-project.org/web/packages/SNPassoc/SNPassoc.pdf\"> <span style=\"font-family:courier new,courier,monospace;\"><strong>SNPassoc</strong> </span></a> package. <br> <br> 
<p><b>Example data sets:</b></p>
<ul>
<li><p><b>REGICOR:</b> These data come from 3 different cross-sectional surveys of individuals representative of the population from a north-west Spanish province (Girona), REGICOR study. It contains 2294 observations on 21 variables such as age, sex, cholesterol profile hypertension, etc.)</p>
</li>
<li><p><b>PREDIMED:</b> The PREDIMED trial (Prevención con Dieta Mediterránea) is a randomized, parallel and multicentric cohort with more than 7,000 participants who were randomly assigned to three diet groups (olive oil + mediterranean diet, nuts + mediterranean diet, and low-fat diet -control group-) and followed-up during more than 7 years.data frame with 6324 observations on the following 15 variables. It contains 6324 observations on 15 variables such as age, sex, diabetes status as well as cardiovascular events during the follow-up time. For more information about the study these data come from, visit http://predimed.onmedic.net</p>
</li>
<li><p><b>SNPs:</b> A data.frame containing 35 selected SNPs and other clinical covariates for 110 cases and 47 controls in a case-control study.</p>
</li>
</ul>
<br><br>
<b>Notes:</b> 
<ul>
<li><p>If you select <b>R-format</b>, the uploaded file should contain a single object of class data.frame.</p></li>
<li><p>If you select <b>TEXT format</b>, make sure to set the right TEXT Options so that your data can be read correctly.</p></li>
<li><p>It may be necessary to refresh the web page when loading a new data set.</p></li>
<li><p>Change the Encoding options to get sure that non-standard characters such as &quot;ñ&quot;, &quot;ç&quot;, etc. are read correctly. </p></li>
</div>
</cghelptext>

SELECT
<cghelptext>
<div style="font-size=25px; text-align:justify">                            
<p><br>Once your data are loaded satisfactory, this panel will be opened automatically.</p>
<p>A list of all variables will apear on the left side list. Use the &quot;<>&quot; button to move the variables from the left side to the right side list and viceversa. Once you have chosen which ones are analysed and which are discarded, press on &quot;Update&quot; button. The remaining variables in the left side will be the ones analysed, i.e. by rows in the descriptive tables.</p>
</div>
</cghelptext>

Type
<cghelptext>
<div style="font-size=25px; text-align:justify">
<p><br>Set the type of your variables to be analysed as:</p>
<ul>
<li><p><strong>Normal: </strong> mean (standard deviation),</p></li> 
<li><p><strong>non-normal:</strong> median [first quantile; third quartile] or as</p></li> 
<li><p><strong>categorical:</strong> absolute frequencies (percentages).</p></li> 
<li><p><strong>NA</strong>: Select this options to test whether a continuous variable is normal or not by a Shapiro test. With this option, a numeric variable with five or less unique different values is also treated as categorical.</p></li>
</ul>
<p>Appropiate statistical tests to compare among groups (if specified) 
are computed depending on the type (normal, non-normal or categorical). </p>
<p>To know more about the tests performed and the functions used in each case see the 
<a href=\"http://www.jstatsoft.org/v57/i12/paper\">published paper in JSS</a>.</p>
<p>Use the &quot;Format&quot; tab from &quot;Step 4. Display&quot; panel to change how statistics are displayed (e.g. to show or not the percentages, etc.)</p>
</div>
</cghelptext>

Response
<cghelptext>
<div style="font-size=25px; text-align:justify">
<p><br>Select which variable indicates the group.</p> 
<p><Three choices are possible:</p>
<ul>
<li><p><strong>None:</strong> Descriptives are performed for the entire data set with no groups.</p></li>
<li><p><strong>Group:</strong> A variable must be selected. By default, only factors variables with five or less categories can be selected. The descriptives will be performed by the groups defined by this variable.</p></li> 
<li><p><strong>Survival:</strong> Use this options to perform survival analysis where the response is a right censored variable, i.e., with time variable and a variable indicating the censoring status (i.e., whether or not the individual suffered the disease). Also, the disease or case status (category) must be indicated.</p></li> 
</ul>
</div>
</cghelptext>

Stratas
<cghelptext>
<div style="font-size=25px; text-align:justify">
<p><br>Select which variable indicates strata.</p> 
<p><Three choices are possible:</p>
<ul>
<li><p><strong>None:</strong> No strata are considered.</p></li>
<li><p><strong>Strata:</strong> A variable must be selected. Only factors variables can be selected. The descriptives tables will be biuld within stratas defined by this variable.</p></li>
</ul>
</div>
</cghelptext>

Hide
<cghelptext>
<div style="font-size=25px; text-align:justify">   
<p><br>Choose the category you want to be hidden for categorical variables.<p>
<p>Additionally, in the &quot;hide no&quot; input text window you can type the category which represents &quot;no&quot; in the sense that the category named as indicated for all binary variables are hidden.</p>
</div>
</cghelptext>

Subset
<cghelptext>
<div style="font-size=25px; text-align:justify"> 
<p><br>Type a selecting criteria to choose which individuals are included.</p>
<p>Also, you can select a subgroup of individuals for each variable. In any case, a logical expression in R language must 
be typed (&quot;==&quot; to compare, &quot;&amp;&quot; to indicate &quot;and&quot;, &quot;&#124;&quot; to indicate &quot;or&quot; and 
&quot;!&quot; to indicate &quot;not&quot;).</p> 
<p>To indicate the category, you must type the number that appears in the &quot;VALUES&quot; table on right side of the application
instead of its name. For example, for gender, type &quot;1&quot; instead of &quot;male&quot; and &quot;2&quot; instead of 
&quot;female&quot;, if &quot;male&quot; is the first category and &quot;female&quot; is the second.</p>
</div>
</cghelptext>

OR/HR
<cghelptext>
<div style="font-size=25px; text-align:justify"> 
<p><br>Set the Odds Ratio (OR) when response is binary or Hazard Ratios (HR) for right censored response.</p>
<p>You can specify the reference category when computing the OR or HR for the categorical variables.
For continuous variables, the scale can be changed. This may be very useful for variables with a wide range such as total cholesterol, 
where it makes more sense to display the OR/HR for each 10 units of change rather than a single unit</p>
</div>
</cghelptext>

Show
<cghelptext>
<div style="font-size=25px; text-align:justify">                         
<p><br>Choose what to be displayed in the bivariate table:</p>
<ul>
<li><p><strong>ALL:</strong> Descriptives for the entire data set.</p></li>
<li><p><strong>p-overall:</strong> p-value comparing the means, medians, proportions or incidence among groups.</p></li> 
<li><p><strong>Descriptives:</strong> Descriptives (means, medians, proportions, etc.) by groups</p></li>
<li><p><strong>p-trend:</strong> p-value for trend. It is supposed that the groups are ordered.</p></li>
<li><p><strong>OR/HR:</strong> Odds Ratio or Hazard Ratio for binary or time-to-event response (grouping variable), respectively.</p></li>
<li><p><strong>Avalaible:</strong> Number of individuals with valid data for each variable and each groups.</p></li>
<li><p><strong>NA category:</strong> For categorical variables, the non-available data is considered as a new category.</p></li>
<li><p><strong>Pairwise p-value:</strong> When more than two groups are considered, p-values corresponding to 2 by 2 comparisons are performed taking into account multiple testing.</p></li>
<li><p><strong>Simplify:</strong> Empty categories (with no available data) are removed from the analyses.</p></li>
</ul>
</div>
</cghelptext>

Format
<cghelptext>
<div style="font-size=25px; text-align:justify">                            
<p><br>Specify how to display the mean and standard deviation, the quantiles and frequencies</p>
</div>
</cghelptext>

Decimals
<cghelptext>
<div style="font-size=25px; text-align:justify"> 
<p><br>Set the number of decimals to be displayed in the bivariate table, for descriptives, p-values and for OR / HR.</p>
</div>
</cghelptext>

Label
<cghelptext>
<div style="font-size=25px; text-align:justify"> 
<p><br>Change the &quot;key&quot; headers of the descriptive table such as &quot;ALL&quot;, &quot;p-value&quot;, etc.</p>
</div>
</cghelptext>

SAVE
<cghelptext>
<div style="font-size=25px; text-align:justify">                           
<p><br>Save the bivariate table in different formats: <b>PDF</b>, <b>CSV</b>, <b>TXT</b>, <b>HTML</b>, <b>Word (.docx)</b> or <b>Excel (.xlsx)</b>.</p>
<p>If <b>PDF</b> format is selected, size can be specified as well as landscape format. These options may be useful when the table 
is big or contains lots of columns.</p>
</div>
</cghelptext>

SUMMARY
<cghelptext>
<div style="font-size=25px; text-align:justify">                            
<p><br>In this tab you can visualize some basic information about the described variables:
<ul>
<li><p><strong>Name:</strong> Name of the variable in the data set.</p></li>
<li><p><strong>Label:</strong> Label of the variable. If the variable has no label (e.g. data imported from a TXT or Excel file), the name of the variable is assigned to the label.</p></li>
<li><p><strong>Method:</strong> Indicates how the variable is treated. By default, character or factor variables are treated as categorical, whereas numeric variables are treated as normal (displaying means and standard deviation) if they follow a normal distribution or non-normal (displaying median and quartiles) otherwise. If a numeric variable contains five or less different unique values, is also treated as categorical and therefore frequencies are displayed.<br> The user can change the default method by using the proper widget in the &quot;Step 3. Setting&quot; panel.</p></li>
<li><p><strong>N:</strong> Number of valid values (i.e. non mising) for each variable.</p></li>
<li><p><strong>Values:</strong> The minimum and maximum value if the variable is numeric or the different categories otherwise. By default a maximum of 10 categories are displayed. Note that a consecutive number is assigned to each category which is used when subsetting. (see info in &quot;Subset&quot; tab inside &quot;Step 3. Settings&quot; panel.</p></li>
</ul>
</div>
</cghelptext>

VALUES
<cghelptext>
<div style="font-size=25px; text-align:justify">                           
<p><br>In this tab the first rows of the data set are displayed.</p>
<p>By making use of DataTable library, it is possible to specify how many rows are visible and navigate through the data set. 
<p>Alo it is possible to order the registries just by clicking on the arrow beside each column name.</p>
<p>Moeover, it is very easy to filter rows by settings the range for numerical variables or levels for categorical variables in the widgets just below each column name.</p>
</div>
</cghelptext>

TABLE
<cghelptext>
<div style="font-size=25px; text-align:justify"> 
<p><br>In this tab the bivariate table is shown.</p>
<p>use the slider input to make it smaller or bigger.</p>
<p>Press the &quot; info &quot; button to display the &quot; info table &quot; containing, for each descripted variable, how many observations are available for each group, whether it is treated as normal, non-normal or categorical, which criteria, if any, has been used to select a subset of individuals and if, the Odds Ratio or Hazard Ratio is computed, the factor for which the variable has been multiplied.</p>
</div>
</cghelptext>

PLOT
<cghelptext>
<div style="font-size=25px; text-align:justify"> 
<p><br>In this tab univariate or bivariate (taking into account the groups if proceeds) plots of selected variable are displayed</p>
<p>Select which variable from the descriptive table you can to plot. Choose between percentages or absolute frequencies (only for categorical variables).</p>
<p></p>You can also save the plot in different formats: <strong>PDF</strong>, <strong>bmp</strong>, <strong>jpg</strong>, <strong>png</strong> and <strong>tiff</strong>.</p>
</div>
</cghelptext>

SNPs
<cghelptext>
<div style="font-size=25px; text-align:justify"> 
<p><br>In this tab descriptives of SNPs (Single Nucleotide Polymorphims) with appropriate statistics and tests for these genetic variants are displayed.</p>
<p>To perform this analysis, only SNPs variables can be selected.</p>
<p>Note that a factor response to display genotype frequencies by groups is permitted but not a time-to-event response variable.</p>
</div>
</cghelptext>

HELPCG
<cghelptext>
<style> shadowtext {display: inline-block;transform: scaleX(1.2);font-size: 140%;color: rgba(250,250,250,0.9);text-shadow: -1px 0 2px rgb(50,50,50), 0 1px 2px rgb(150,150,150), 1px 0 2px rgb(200,200,200), 0 -1px 2px rgb(250,250,250);font-family: 'Courier New', Courier, 'Lucida Sans Typewriter', 'Lucida Typewriter', monospace;}</style>
<div style="color: grey; font-family: arial, helvetica, sans-serif;background-color:rgb(204,214,235); color:#385188; border-radius: 8px; padding:20px; border:2px solid #385188; font-size:17px; margin-bottom:2%; margin-top:2%; margin-right:0%;margin-left:0%;line-height:130%;box-shadow: 6px 6px 5px grey; text-align:justify">
<p><strong><span style="color:#0099ff;"><font size="2px"><span style="font-size:22px;"><span style="font-family:courier new,courier,monospace;">compare</span></span></font></span><shadowtext>~</shadowtext><span style="color:#0000cc;"><font size="2px"><span style="font-size:22px;"><span style="font-family:courier new,courier,monospace;">Groups</span></span></font></span></strong><span style="font-size:22px;"> <span style="font-family:arial,helvetica,sans-serif;"><strong>Project</strong></span></span></p>
<hr style="border: 1px solid grey;">
<ul style="margin-left:-1%">
<li><p style="font-size: 16px;">Create data summaries for quality control, extensive reports for exploring data, as well as <strong>publication-ready univariate or bivariate tables</strong> in several formats:</p>
<ul>
<li><p><strong>plain text (csv)</strong>, <strong>HTML</strong>, <strong>PDF</strong>, <strong>Word</strong> or <strong>Excel</strong></p></li>
</ul>
</li>
<li>
<p style="font-size: 16px;">
<strong>Perform figures</strong> to quickly visualise the distribution of your data (boxplots, barplots, normality-plots, etc.).
<li>
<p style="font-size: 16px;">
<strong>Display statistics</strong> (mean, median, frequencies, incidences, etc.)</p>
</li>
<li>
<p style="font-size: 16px;">
<strong>Perform the appropriate tests</strong> (t-test, Analysis of variance, Kruskal-Wallis, Fisher, log-rank, ...) depending on the nature of the described variable (normal, non-normal or qualitative).</p>
</li>
<li>
<p style="font-size: 16px;">
<strong>Summarize genetic data</strong> (Single Nucleotids Polymorphisms) data displaying Allele Frequencies and performing Hardy-Weinberg Equilibrium tests among other typical statistics and tests for these kind of data.
</p>
</li>
</ul>  
</div>
</cghelptext>

HELPWUI
<cghelptext>
<style> shadowtext {display: inline-block;transform: scaleX(1.2);font-size: 140%;color: rgba(250,250,250,0.9);text-shadow: -1px 0 2px rgb(50,50,50), 0 1px 2px rgb(150,150,150), 1px 0 2px rgb(200,200,200), 0 -1px 2px rgb(250,250,250);font-family: 'Courier New', Courier, 'Lucida Sans Typewriter', 'Lucida Typewriter', monospace;}</style>
<br><hr style="display: block; height: 1px;border: 0; border-top: 1px solid #ccc; margin: 1em 0; padding: 0;">
<div style="color: grey; font-family: arial, helvetica, sans-serif; text-align:justify">
<p><strong><span style="color:#0099ff;"><font size="2px"><span style="font-size:22px;"><span style="font-family:courier new,courier,monospace;">compare</span></span></font></span><shadowtext>~</shadowtext><span style="color:#0000cc;"><font size="2px"><span style="font-size:22px;"><span style="font-family:courier new,courier,monospace;">Groups</span></span></font></span></strong><font color="grey" size="2px"><span style="font-size:22px;"> <span style="font-family:arial,helvetica,sans-serif;"><strong>Web User Interface</strong> (WUI)</span></span></font></p>
<p style="font-size: 16px;">This WUI is designed to allow clinicians and scientists to take advantage of the functionality of <span style="font-family:courier new,courier,monospace;"><strong>compareGroups</strong></span>&nbsp;to analyse their epidemiological data, but without the need to install or learn how to use <span style="font-family:courier new,courier,monospace;">R</span>.</p>
<p style="font-size: 16px;">This system has all of the functionality of the current command line version of compareGroups, but has the advantage of allowing you to load and analyse your data on any device with an internet connection.</p>
<p style="font-size: 16px;">This WUI has been build entirely using <a href="http://shiny.rstudio.com/">Shiny</a> tools.</p>
</div>
</cghelptext>

DATASECURITY
<cghelptext>
<style> shadowtext {display: inline-block;transform: scaleX(1.2);font-size: 140%;color: rgba(250,250,250,0.9);text-shadow: -1px 0 2px rgb(50,50,50), 0 1px 2px rgb(150,150,150), 1px 0 2px rgb(200,200,200), 0 -1px 2px rgb(250,250,250);font-family: 'Courier New', Courier, 'Lucida Sans Typewriter', 'Lucida Typewriter', monospace;}</style>
<br><hr style="display: block; height: 1px;border: 0; border-top: 1px solid #ccc; margin: 1em 0; padding: 0;">
<div style="color: grey; font-family: arial, helvetica, sans-serif; text-align:justify">
<p><span style="color: grey; font-family: arial, helvetica, sans-serif; font-size: 22px;"><strong>Data Security</strong> when using the </span><strong><span style="color:#0099ff;"><font size="2px"><span style="font-size:22px;"><span style="font-family:courier new,courier,monospace;">compare</span></span></font></span><shadowtext>~</shadowtext><span style="color:#0000cc;"><font size="2px"><span style="font-size:22px;"><span style="font-family:courier new,courier,monospace;">Groups</span></span></font></span></strong><span style="color: grey; font-family: arial, helvetica, sans-serif; font-size: 22px;">&nbsp;WUI</span></p>
<p style="font-size: 16px;"><span style="font-family:courier new,courier,monospace;"><strong>compareGroups</strong></span> is developed by experienced researchers who are actively involved in epidemiological research on human subjects. We understand very well how important it is to guarantee data confidentiality and security, and we have designed the compareGroups WUI so that your data will remain as secure as at your own institute.</p>
<p style="font-size: 16px;"><strong>Will my data be safe if I load them into the compareGroups WUI?</strong><br>
The compareGroups WUI is hosted on&nbsp;a <a href="https://www.digitalocean.com/security/" target="_blank">secure cloud server platform</a>, which guarantees the physical and electronic security of the WUI, and therefore of any data that are loaded into it. Data loaded into the WUI are transmitted over a secure connection via HTTPS.</p>
<p style="font-size: 16px;"><strong>What happens to my data when I load them into the compareGroups WUI?</strong><br>
If you&#39;re familiar with R, you&#39;ll know that it loads all data and performs all calculations within the&nbsp; RAM&nbsp; allocated for that R session. This means that it never writes data to the disk, and when the R session ends, all data it has been working on are eliminated.</p>
<p style="font-size: 16px;">Accessing the <span style="font-family:courier new,courier,monospace;"><strong>compareGroups</strong></span> <strong>WUI</strong> page starts a dedicated session of R to read and manipulate your data, and compute your results. However, this R session never saves data to the hard disk of the server it&#39;s running on, and only sends results to the WUI, either in the form of displayed tables or plots, or as files to be downloaded by the the user via the SAVE panel.</p>
<p style="font-size: 16px;">When you close your browser window, the R session is terminated, and all data are lost. To can verify this for yourself by simply re-loading the WUI page, after which you will see that no data are loaded.&nbsp; <span style="font-family:courier new,courier,monospace;"><strong>compareGroups</strong></span> developers&nbsp; only&nbsp;have access to the underlying R code that generates that WUI and performes the calculations, and can never gain access to data that have been loaded into a remote session.</p>
<p style="font-size: 16px;">If you have any further concerns or questions about the security of your data when using the compareGroups WUI, please get in touch with us via the <a href="http://www.comparegroups.eu/contact">contact form</a>.</p>
</div>
</cghelptext>
